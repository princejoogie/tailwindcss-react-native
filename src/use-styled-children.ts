import { ReactNode, Children, cloneElement } from "react";
import { isFragment } from "react-is";
import { matchChildAtRule } from "./match-at-rule";
import { AtRuleRecord } from "./types/common";

export interface UseStyledChildrenOptions {
  componentChildren: ReactNode;
  childStyles?: AtRuleRecord[];
}

export function useStyledChildren({
  componentChildren,
  childStyles,
}: UseStyledChildrenOptions): ReactNode {
  let children = isFragment(componentChildren)
    ? // This probably needs to be recursive
      componentChildren.props.children
    : componentChildren;

  if (childStyles) {
    children = Children.map(children, (child, index) => {
      const matchingStyles = [];
      for (const { atRules, ...styles } of childStyles) {
        const matches = atRules.every(([rule, params]) => {
          return matchChildAtRule({
            nthChild: index + 1,
            rule,
            params,
          });
        });
        if (matches) {
          matchingStyles.push(styles);
        }
      }

      return cloneElement(child, {
        style: child.props.style
          ? [child.props.style, matchingStyles]
          : matchingStyles.length > 0
          ? matchingStyles
          : undefined,
      });
    });
  }

  return children;
}
