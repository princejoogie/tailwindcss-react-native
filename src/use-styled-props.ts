import { StyleProp } from "react-native";
import { UseTailwindCallback } from "./use-tailwind";
import { ChildClassNameSymbol } from "./utils/child-styles";

export interface UseStyledPropsOptions {
  tw: UseTailwindCallback<any>;
  classes: string | undefined;
  componentStyles: StyleProp<any>;
  propsToTransform?: boolean | string[];
  componentProps: Record<string, unknown>;
}

export function useStyledProps({
  tw,
  classes,
  componentStyles,
  propsToTransform,
  componentProps,
}: UseStyledPropsOptions) {
  const mainStyles = tw(classes);

  const style = componentStyles ? [mainStyles, componentStyles] : mainStyles;

  const styledProps: Record<string, unknown> = {};

  if (propsToTransform) {
    if (propsToTransform === true) {
      propsToTransform = Object.keys(componentProps);
    }

    for (const prop of propsToTransform) {
      const value = componentProps[prop];

      if (typeof value === "string") {
        styledProps[prop] = tw(value);
      }
    }
  }

  console.log(styledProps);

  return {
    childStyles: mainStyles[ChildClassNameSymbol],
    style,
    ...styledProps,
  };
}
