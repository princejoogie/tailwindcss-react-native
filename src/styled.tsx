import {
  createElement,
  FunctionComponent,
  ComponentClass,
  PropsWithChildren,
  ComponentProps,
} from "react";
import { ImageStyle, StyleProp, TextStyle, ViewStyle } from "react-native";
import { useTailwind } from "./use-tailwind";
import { useInteraction } from "./use-interaction";
import { ComponentContext } from "./context";
import { useStyledProps } from "./use-styled-props";
import { useStyledChildren } from "./use-styled-children";

type StyledProps<P> = PropsWithChildren<
  P & {
    className?: string;
    tw?: string;
    style?: StyleProp<ViewStyle | TextStyle | ImageStyle>;
  }
>;

type Component<P> = string | FunctionComponent<P> | ComponentClass<P>;

export interface StyledOptions {
  props?: boolean | string[];
}

export function styled<P>(
  Component: Component<P>,
  { props: propsToTransform = false }: StyledOptions = {}
): FunctionComponent<StyledProps<P>> {
  function Styled({
    className,
    tw,
    style: componentStyles,
    children: componentChildren,
    ...componentProps
  }: StyledProps<P>) {
    const { hover, focus, active, ...handlers } =
      useInteraction(componentProps);

    const classes = tw ?? className ?? "";

    const twCallback = useTailwind({
      hover,
      focus,
      active,
      flatten: false,
    });

    const { childStyles, ...styledProps } = useStyledProps({
      tw: twCallback,
      classes,
      componentStyles,
      propsToTransform,
      componentProps,
    });

    const children = useStyledChildren({
      componentChildren,
      childStyles,
    });

    const element = createElement(Component, {
      ...componentProps,
      ...handlers,
      ...styledProps,
      children,
    } as unknown as P);

    return !classes.split(/\s+/).includes("component")
      ? element
      : createElement<ComponentProps<typeof ComponentContext.Provider>>(
          ComponentContext.Provider,
          {
            children: element,
            value: { hover, focus, active },
          }
        );
  }

  if (typeof Component !== "string") {
    Styled.displayName = `TailwindCssReactNative.${
      Component.displayName || Component.name || "NoName"
    }`;
  }

  return Styled;
}

type StyledComponentProps<P> = StyledProps<P> & {
  component: Component<P>;
};

export function StyledComponent<P>({
  component,
  ...options
}: StyledComponentProps<P>) {
  const Component = styled<P>(component);
  return <Component {...(options as P)} />;
}
