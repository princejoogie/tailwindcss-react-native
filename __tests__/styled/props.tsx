import { render } from "@testing-library/react-native";
import { View, ViewProps, ViewStyle } from "react-native";
import { styled } from "../../src";
import { TestProvider } from "../tailwindcss/runner";

const StyledTestComponent = styled(
  ({ style, style2 }: ViewProps & { style2: ViewStyle }) => {
    return (
      <>
        <View style={style} />
        <View style={style2} />
      </>
    );
  },
  {
    props: ["style2"],
  }
) as any;

describe("Styled - Custom Props", () => {
  test("can style custom props", () => {
    const tree = render(
      <TestProvider css="m-1 p-2">
        <StyledTestComponent className="m-1" style2="p-2" />
      </TestProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
