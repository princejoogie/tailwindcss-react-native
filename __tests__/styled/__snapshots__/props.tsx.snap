// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styled - Custom Props can style custom props 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "marginBottom": 4,
          "marginLeft": 4,
          "marginRight": 4,
          "marginTop": 4,
        },
      ]
    }
  />,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
  />,
]
`;
